### PR TITLE
8320144: Compilation crashes when a custom annotation with invalid default value is used

### PR DESCRIPTION
**Pull request by biboudis**
The erroneous case of:

`public String[] excludeModules() default new String[0];`

was not short-circuited in the compiler and we were getting an NPE in `l.nonEmpty()` for `l`. Now, `getAnnotationArrayValue` returns the error attribute, if that case is detected.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320144](https://bugs.openjdk.org/browse/JDK-8320144): Compilation crashes when a custom annotation with invalid default value is used (**Bug** - P4)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [20c3ca05](https://git.openjdk.org/jdk/pull/16977/files/20c3ca050c492c55c7981220293e16bd254f814e)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16977/head:pull/16977` \
`$ git checkout pull/16977`

Update a local copy of the PR: \
`$ git checkout pull/16977` \
`$ git pull https://git.openjdk.org/jdk.git pull/16977/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16977`

View PR using the GUI difftool: \
`$ git pr show -t 16977`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16977.diff">https://git.openjdk.org/jdk/pull/16977.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16977#issuecomment-1841178066)

### CODE REVIEW

File: test/langtools/tools/javac/T8320144.out [hunk 1]
```diff
@@ -0,0 +1,2 @@
+T8320144.java:16:54: compiler.err.new.not.allowed.in.annotation
+1 error
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/T8320144.out:2]: there seems to be an empty new line missing at the end of the file

**Review by vicente-romero-oracle — APPROVED** [review]
looks sensible to me

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Annotate.java [hunk 2]
```diff
@@ -710,16 +709,18 @@ private Attribute getAnnotationArrayValue(Type expectedElementType, JCExpression
         JCNewArray na = (JCNewArray)tree;
         if (na.elemtype != null) {
             log.error(na.elemtype.pos(), Errors.NewNotAllowedInAnnotation);
+            return new Attribute.Error(syms.errType);
+        } else {
+            ListBuffer<Attribute> buf = new ListBuffer<>();
+            for (List<JCExpression> l = na.elems; l.nonEmpty(); l = l.tail) {
+                buf.append(attributeAnnotationValue(types.elemtype(expectedElementType),
```
Comment (lahodaj) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Annotate.java:716]: Here, I would be more inclined to attribute the values even in the erroneous case. Like if someone writes:
```
public String[] value() default new String[0] {Other.CONSTANT}
```

Attributing even for the erroneous case would mean the AST would have types/symbols filled in, for possible further use.

**Review by lahodaj — APPROVED** [review]
Looks great to me.

**Pull request closed without merging**